### DISCUSSION TITLE
Is there a way to daynamic rewrite subscription topic?

### DISCUSSION THREAD

**Discussion by tubzby** [emqx/emqx#10585] [Q&A]
For the following workflow:
1. User subscribes to "user/"
2. EMQX forward the request to my Backend server.
3. Backend server verified the request, and sends an array of topics to EMQX, such as 
   a) "user/USERID": messages sent to this user
   b) "user/GROUPID": messages sent to this group
   c) "broadcast/": broadcast to all users

The reason why I don't want the user to directly subscribe to those topics is that I want to hide the details, and maybe these topics will be removed or replaced for later releases.

I have checked out auto-subscription and proxy-subscription,  it sounds like what I need, but can EMQX proxy the request to my backend server, because the USERID/GROUPID is stored in the database?

**Comment by SergeTupchiy** [comment]
Hi @tubzby,

Have you checked out the available topic rewrite functionality? 
https://www.emqx.io/docs/en/v5.0/mqtt/mqtt-topic-rewrite.html#configure-topic-rewrite-rules
https://docs.emqx.com/en/enterprise/v4.4/modules/topic_rewrite.html#topic-rewrite

It supports client id / username placeholders (`%c`, `%u` in EMQX 4.4 and `${clientid}`, `${username}` in EMQX 5.x),
so a client can subscribe to `user` and EMQX can rewrite it to `user/${username}`.

**Comment by SergeTupchiy** [comment]
We also have more advanced features like plugins and gRPC hooks:
https://docs.emqx.com/en/enterprise/v5.0/dashboard/extensions.html#exhook
https://docs.emqx.com/en/enterprise/v5.0/dashboard/extensions.html#plugins
https://docs.emqx.com/en/enterprise/v5.0/extensions/hooks.html#callback-functions-chain

It may be possible to use `client.subscribe` hook to fetch and update TopicFilters, but it would require some development efforts. 
Here is a plugin template: https://github.com/emqx/emqx-plugin-template/blob/master/src/emqx_plugin_template.erl#L119

**Comment by tubzby** [comment]
@SergeTupchiy Thanks for your suggestion, gRPC seems to match what I need. 
I checked the function signature at [here](https://www.emqx.io/docs/en/v5.0/advanced/lang-exhook.html#apis):
```
  rpc OnClientSubscribe(ClientSubscribeRequest) returns (EmptySuccess) {};
```

Does this mean it can only return an empty success instead of an array of TopicFilters?

**Reply by SergeTupchiy** [reply]
@tubzby, 
You are right, for now gRPC hook indeed doesn't allow to return any data `OnClientSubscribe`. Apologies for the confusion.
It still must be possible to rewrite topics on subscribe with the plugin, though.
